### ISSUE TITLE
Change port 443 for nested edge scenario

### ISSUE THREAD

**Issue by mramin2000** [Azure/iotedge#5612]
Hello,

I'm working with the nested edge scenario and I'd like to use port 443 on both the child and the parent device for different service, but it seems like the devices are communicating on port 443 among themselves. Is it possible to change this port?
I've already changed the portbindings for edgeHub and the API Proxy module on another port, but this didn't help.

**Comment by huguesBouvier** [comment]
@dmolokanov Yes, the process is described here: https://docs.microsoft.com/en-us/azure/iot-edge/how-to-configure-api-proxy-module?view=iotedge-2020-11#minimize-open-ports.
The changes need to be made on the parent.

**Comment by mramin2000** [comment]
Now it works. Thanks.

**Comment by huguesBouvier** [comment]
ahh! Just realized: Did you put back the API proxy from 443 to 8000?
Then you need to put edgeHub back on 443. That is why you are seeing this issue.
You can't use 443 without the API proxy, it is used by edgeHub.

API proxy is configurable, if your applications are using HTTP, it should be possible to configure the API proxy to route your messages. 
Please see:
Readme: https://github.com/Azure/iotedge/tree/master/edge-modules/api-proxy-module
API proxy config file: https://github.com/Azure/iotedge/blob/master/edge-modules/api-proxy-module/templates/nginx_default_config.conf

**Comment by huguesBouvier** [comment]
Hum, I never used that option, if it doesn't work, I see only this:
- Use the API proxy, that's what I would recommend
- Map 443 to 8000 or 8443 on the edgeHub parent and use the routing table on the children
- Add a reverse proxy between child and parent to reroute messages (can also be done locally)

**Comment by huguesBouvier** [comment]
Looking hard at it, it is difficult to pinpoint what the issue could be. 
I can't really help for that level of details. If IPtable doesn't work for you, what protocol is your application using? If it uses http, then loading your own config in api should be easier.

**Comment by huguesBouvier** [comment]
Can I close this issue? I can provide help for configuring IP proxy but it will not be possible for IP tables.
Thanks!

**Comment by mramin2000** [comment]
Our application uses https, so we need incoming traffic on port 443 for our service. Looking at the nginx config from [here](https://github.com/Azure/iotedge/blob/master/edge-modules/api-proxy-module/templates/nginx_default_config.conf) I found that the edgeHub endpoint is mapped to https as default. Could it help to change this part (line 17,18) to another port like 8443? If not, do you have any suggestions for what I have to change in the API Proxy config? We need to reroute edgeHub traffic to another port like 8443. And thanks a lot for your help so far!

**Comment by huguesBouvier** [comment]
You can change the configuration very easily to 8443 by adding the env var: NGINX_DEFAULT_PORT=8443 but it wont help you because you will also need to change the downstream edgeHub.

**However, the API proxy can handle 2 connection on the same port and route them to different target.**

For example look at the piece of code in the config:
```


        **if ($uri ~ ^/v2*.)**
        {
            **rewrite ^(.*)$ /registry$1 last;**
        }

        **location ~^/registry/(.*) {**
            proxy_http_version 1.1;
            resolver 127.0.0.11;
            set $endpoint http://${DOCKER_REQUEST_ROUTE_ADDRESS}/$1$is_args$args;
            proxy_pass $endpoint;
        }

        **if ($http_x_ms_version)**
        {
            **rewrite ^(.*)$ /storage$1 last;**
        }

        **location ~^/storage/(.*){**
            resolver 127.0.0.11;
            proxy_http_version 1.1;
            set $endpoint http://${BLOB_UPLOAD_ROUTE_ADDRESS}/$1$is_args$args;
            proxy_pass $endpoint;
        }
```

if the request contains the header "http_x_ms_version" it is routed to the blob storage module
if the request uri start with "/v2" then it is routed to the "registry module" (for docker)

You can make your own rule:
if the request come from your custom module below, route it to your custom module upstream.

For debugging I suggest that you mount the file /app/nginx_default_config.conf on your drive so you can make changes fast.
do nginx -s stop in api proxy container to reload nginx with the new config.

**Comment by mramin2000** [comment]
I tried it again with iptables and made it finally working. It was much easier than I thought in the other tries.
On the parent I have deleted the 443/443 port binding for edgeHub and set API Proxy on Nginx Port 443 with port binding for 443 to hostport 8443. On the child I set edgeHub Portbinding on 443 to Hostport 8443 and put in $upstream:443 for the modules.
Additionally I made one rule on the parent: iptables -t nat -A PREROUTING -p tcp -s <childIP> --dport 443 -j REDIRECT --to-ports 8443

Thanks for your help!
